File: backend/llm/provider.py
"""Адаптер языковой модели. Смена провайдера правит только этот файл
(docs/04-decisions.md, решение №5)."""
import json
from dataclasses import dataclass

from backend import config
from backend.models import LEVELS


class LLMUnavailable(RuntimeError):
    """Модель не ответила или ответила не по схеме. Обрабатывается повтором."""


@dataclass(frozen=True)
class AnalysisRequest:
    criteria: str
    sender: str
    subject: str
    body: str
    source: str
    overrides: list[tuple[str, str]]  # (тема, уровень, который поставил пользователь)


@dataclass(frozen=True)
class AnalysisResult:
    level: str
    category: str
    deadline: str
    needs_reply: bool
    needs_action: bool
    summary: str


RESPONSE_SCHEMA = {
    "type": "object",
    "additionalProperties": False,
    "required": ["level", "category", "deadline", "needs_reply", "needs_action", "summary"],
    "properties": {
        "level": {"type": "string", "enum": list(LEVELS)},
        "category": {"type": "string"},
        "deadline": {"type": "string"},
        "needs_reply": {"type": "boolean"},
        "needs_action": {"type": "boolean"},
        "summary": {"type": "string"},
    },
}

SYSTEM_PROMPT = (
    "Ты помощник, который сортирует входящие сообщения по важности для одного человека. "
    "Отвечай строго JSON по заданной схеме. Все текстовые поля — на русском языке, "
    "независимо от языка исходного сообщения.\n"
    "level — один из: CRITICAL (нельзя пропустить, есть жёсткий срок или риск), "
    "HIGH (важно, требует внимания сегодня-завтра), NORMAL (обычное рабочее), "
    "LOW (рассылки, уведомления сервисов, ничего не требуют).\n"
    "category — 1–2 слова, например «Финансы», «Юридическое», «Найм». "
    "Если тема непонятна, верни пустую строку.\n"
    "deadline — срок словами из сообщения: «Сегодня, 18:00», «Пятница». "
    "Если срока нет — пустая строка.\n"
    "summary — 1–2 предложения о сути сообщения."
)


def build_user_prompt(request: AnalysisRequest) -> str:
    parts = [
        "Критерии важности пользователя (свободный текст, не жёсткие правила):",
        request.criteria.strip() or "— не заданы, оценивай на общих основаниях —",
        "",
        f"Источник: {request.source}",
        f"Отправитель: {request.sender}",
        f"Тема: {request.subject}",
        "Текст:",
        request.body[: config.LLM_MAX_BODY_CHARS],
    ]
    if request.overrides:
        parts += [
            "",
            "Ручные исправления пользователя по прошлым сообщениям "
            "(тема → уровень, который он считает верным):",
        ]
        parts += [f"- {subject} → {level}" for subject, level in request.overrides]
    return "\n".join(parts)


def _client():
    if not config.OPENAI_API_KEY:
        raise LLMUnavailable("OPENAI_API_KEY не задан")
    from openai import OpenAI  # импорт здесь: без ключа пакет не нужен

    return OpenAI(api_key=config.OPENAI_API_KEY)


def analyze(request: AnalysisRequest) -> AnalysisResult:
    """Один вызов модели на сообщение. Любой сбой — LLMUnavailable."""
    try:
        response = _client().chat.completions.create(
            model=config.OPENAI_MODEL,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": build_user_prompt(request)},
            ],
            response_format={
                "type": "json_schema",
                "json_schema": {"name": "analysis", "schema": RESPONSE_SCHEMA, "strict": True},
            },
            temperature=0,
        )
        raw = response.choices[0].message.content or ""
    except LLMUnavailable:
        raise
    except Exception as error:  # сеть, лимиты, отказ провайдера
        raise LLMUnavailable(str(error)) from error
    return parse_response(raw)


def parse_response(raw: str) -> AnalysisResult:
    """Разбор ответа модели. Всё, что не соответствует схеме, — ошибка вызова
    (docs/00-product-spec.md, п. 6.2)."""
    try:
        data = json.loads(raw)
    except (json.JSONDecodeError, TypeError) as error:
        raise LLMUnavailable(f"ответ модели не разобрать: {raw[:200]}") from error
    if not isinstance(data, dict):
        raise LLMUnavailable("ответ модели не объект")
    level = data.get("level")
    if level not in LEVELS:
        raise LLMUnavailable(f"неизвестный уровень: {level!r}")
    for key in ("category", "deadline", "summary"):
        if not isinstance(data.get(key), str):
            raise LLMUnavailable(f"поле {key} не строка")
    for key in ("needs_reply", "needs_action"):
        if not isinstance(data.get(key), bool):
            raise LLMUnavailable(f"поле {key} не булево")
    return AnalysisResult(
        level=level,
        category=data["category"].strip(),
        deadline=data["deadline"].strip(),
        needs_reply=data["needs_reply"],
        needs_action=data["needs_action"],
        summary=data["summary"].strip(),
    )

File: backend/internal/openai/provider.go
// Package llm — адаптер языковой модели. Смена провайдера правит только этот
// файл (docs/04-decisions.md, решение №5).
package openai

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"personalinbox/internal/exceptions"
	"strings"
	"time"

	"personalinbox/internal/core"
	"personalinbox/internal/sqlite"
)

// Request — всё, что уходит в модель об одном сообщении.
type Request struct {
	Criteria  string
	Sender    string
	Subject   string
	Body      string
	Source    string
	Overrides []sqlite.Override
}

// Result — строгий JSON от модели, разобранный в структуру.
type Result struct {
	Level       string
	Category    string
	Deadline    string
	NeedsReply  bool
	NeedsAction bool
	Summary     string
}

// Analyzer — то, что умеет оценить сообщение. Тесты подставляют свою реализацию.
type Analyzer interface {
	Analyze(Request) (Result, error)
}

const systemPrompt = "Ты помощник, который сортирует входящие сообщения по важности для одного человека. " +
	"Отвечай строго JSON по заданной схеме. Все текстовые поля — на русском языке, " +
	"независимо от языка исходного сообщения.\n" +
	"level — один из: CRITICAL (нельзя пропустить, есть жёсткий срок или риск), " +
	"HIGH (важно, требует внимания сегодня-завтра), NORMAL (обычное рабочее), " +
	"LOW (рассылки, уведомления сервисов, ничего не требуют).\n" +
	"category — 1–2 слова, например «Финансы», «Юридическое», «Найм». " +
	"Если тема непонятна, верни пустую строку.\n" +
	"deadline — срок словами из сообщения: «Сегодня, 18:00», «Пятница». " +
	"Если срока нет — пустая строка.\n" +
	"summary — 1–2 предложения о сути сообщения."

// responseSchema — structured output: всё, что не соответствует схеме,
// считается ошибкой вызова.
func responseSchema() map[string]any {
	return map[string]any{
		"type":                 "object",
		"additionalProperties": false,
		"required": []string{
			"level", "category", "deadline", "needs_reply", "needs_action", "summary",
		},
		"properties": map[string]any{
			"level":        map[string]any{"type": "string", "enum": sqlite.Levels},
			"category":     map[string]any{"type": "string"},
			"deadline":     map[string]any{"type": "string"},
			"needs_reply":  map[string]any{"type": "boolean"},
			"needs_action": map[string]any{"type": "boolean"},
			"summary":      map[string]any{"type": "string"},
		},
	}
}

// BuildUserPrompt собирает пользовательскую часть промпта: критерии, само
// сообщение и последние ручные исправления как обратную связь.
func BuildUserPrompt(request Request) string {
	criteria := strings.TrimSpace(request.Criteria)
	if criteria == "" {
		criteria = "— не заданы, оценивай на общих основаниях —"
	}
	body := request.Body
	if len([]rune(body)) > core.LLMMaxBodyChars {
		body = string([]rune(body)[:core.LLMMaxBodyChars])
	}
	parts := []string{
		"Критерии важности пользователя (свободный текст, не жёсткие правила):",
		criteria,
		"",
		"Источник: " + request.Source,
		"Отправитель: " + request.Sender,
		"Тема: " + request.Subject,
		"Текст:",
		body,
	}
	if len(request.Overrides) > 0 {
		parts = append(parts, "",
			"Ручные исправления пользователя по прошлым сообщениям "+
				"(тема → уровень, который он считает верным):")
		for _, override := range request.Overrides {
			parts = append(parts, fmt.Sprintf("- %s → %s", override.Subject, override.Level))
		}
	}
	return strings.Join(parts, "\n")
}

// OpenAI — вызов chat completions напрямую по HTTP: SDK ради одного запроса
// избыточен, а зависимостей у pet-проекта должно быть меньше.
type OpenAI struct {
	APIKey string
	Model  string
	// BaseURL позволяет подменить адрес в тестах.
	BaseURL string
	Client  *http.Client
}

// NewOpenAI собирает адаптер из настроек.
func NewOpenAI(cfg core.Config) *OpenAI {
	return &OpenAI{
		APIKey:  cfg.OpenAIKey,
		Model:   cfg.OpenAIModel,
		BaseURL: "https://api.openai.com/v1",
		Client:  &http.Client{Timeout: 60 * time.Second},
	}
}

// Analyze — один вызов модели на сообщение. Любой сбой — exceptions.ErrModelUnavailable.
func (o *OpenAI) Analyze(request Request) (Result, error) {
	if o.APIKey == "" {
		return Result{}, fmt.Errorf("%w: OPENAI_API_KEY не задан", exceptions.ErrModelUnavailable)
	}
	payload := map[string]any{
		"model": o.Model,
		"messages": []map[string]string{
			{"role": "system", "content": systemPrompt},
			{"role": "user", "content": BuildUserPrompt(request)},
		},
		"response_format": map[string]any{
			"type": "json_schema",
			"json_schema": map[string]any{
				"name":   "analysis",
				"schema": responseSchema(),
				"strict": true,
			},
		},
		"temperature": 0,
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return Result{}, fmt.Errorf("%w: %v", exceptions.ErrModelUnavailable, err)
	}

	base := o.BaseURL
	if base == "" {
		base = "https://api.openai.com/v1"
	}
	httpRequest, err := http.NewRequest(http.MethodPost, base+"/chat/completions", bytes.NewReader(body))
	if err != nil {
		return Result{}, fmt.Errorf("%w: %v", exceptions.ErrModelUnavailable, err)
	}
	httpRequest.Header.Set("Content-Type", "application/json")
	httpRequest.Header.Set("Authorization", "Bearer "+o.APIKey)

	client := o.Client
	if client == nil {
		client = &http.Client{Timeout: 60 * time.Second}
	}
	response, err := client.Do(httpRequest)
	if err != nil {
		return Result{}, fmt.Errorf("%w: %v", exceptions.ErrModelUnavailable, err)
	}
	defer response.Body.Close()

	raw, err := io.ReadAll(response.Body)
	if err != nil {
		return Result{}, fmt.Errorf("%w: %v", exceptions.ErrModelUnavailable, err)
	}
	if response.StatusCode != http.StatusOK {
		return Result{}, fmt.Errorf("%w: провайдер ответил %d", exceptions.ErrModelUnavailable, response.StatusCode)
	}

	var completion struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.Unmarshal(raw, &completion); err != nil || len(completion.Choices) == 0 {
		return Result{}, fmt.Errorf("%w: ответ провайдера не разобрать", exceptions.ErrModelUnavailable)
	}
	return ParseResponse(completion.Choices[0].Message.Content)
}

// ParseResponse разбирает ответ модели. Всё, что не соответствует схеме, —
// ошибка вызова (docs/00-product-spec.md, п. 6.2).
func ParseResponse(raw string) (Result, error) {
	var data map[string]any
	if err := json.Unmarshal([]byte(raw), &data); err != nil {
		short := raw
		if len(short) > 200 {
			short = short[:200]
		}
		return Result{}, fmt.Errorf("%w: ответ модели не разобрать: %s", exceptions.ErrModelUnavailable, short)
	}

	level, _ := data["level"].(string)
	if !sqlite.Contains(sqlite.Levels, level) {
		return Result{}, fmt.Errorf("%w: неизвестный уровень: %v", exceptions.ErrModelUnavailable, data["level"])
	}
	text := map[string]string{}
	for _, key := range []string{"category", "deadline", "summary"} {
		value, ok := data[key].(string)
		if !ok {
			return Result{}, fmt.Errorf("%w: поле %s не строка", exceptions.ErrModelUnavailable, key)
		}
		text[key] = strings.TrimSpace(value)
	}
	flags := map[string]bool{}
	for _, key := range []string{"needs_reply", "needs_action"} {
		value, ok := data[key].(bool)
		if !ok {
			return Result{}, fmt.Errorf("%w: поле %s не булево", exceptions.ErrModelUnavailable, key)
		}
		flags[key] = value
	}
	return Result{
		Level:       level,
		Category:    text["category"],
		Deadline:    text["deadline"],
		NeedsReply:  flags["needs_reply"],
		NeedsAction: flags["needs_action"],
		Summary:     text["summary"],
	}, nil
}
